src/Core/Material/MaterialModel.cpp

```diff
@@ -8,6 +9,71 @@ void MaterialModel::displayInfo() const {
     using namespace Core::Utils; // log
     LOG( logERROR ) << "MaterialModel : unknown material type : " << m_materialType;
 }
+
+Utils::Color MaterialModel::evalBSDF( Vector3f w_i, Vector3f w_o, Vector3f normal, Vector2f uv ) {
+    return Utils::Color();
+}
+
+std::optional<std::pair<Vector3f, Scalar>> MaterialModel::sample( Vector3f inDir, Vector3f normal, Vector2f u ) {
+    return {};
+}
+
+Scalar MaterialModel::PDF( Vector3f inDir, Vector3f outDir, Vector3f normal ) {
+    return 0.0f;
+}
+
+void MaterialModel::coordinateSystem( Vector3f normal, Vector3f* tangent, Vector3f* bitangent ) {
+    if (std::abs(normal[0]) > std::abs(normal[1])) {
+        Scalar invLen = 1 / std::sqrt(normal[0] * normal[0] + normal[2] * normal[2]);
+        *tangent = Vector3f(-normal[2] * invLen, 0, normal[0] * invLen);
+    } else {
+        Scalar invLen = 1 / std::sqrt(normal[1] * normal[1] + normal[2] * normal[2]);
+        *tangent = Vector3f(0, normal[2] * invLen, -normal[1] * invLen);
+    }
+    *bitangent = normal.cross(*tangent);
+}
+
+Vector3f MaterialModel::sampleSpecular(Vector3f inDir, Vector2f u) {
+    Vector3f nextDir, halfway;
+
+    halfway = sampleHemisphereCosineWeighted(u);
+    nextDir = inDir - 2.0f * inDir.dot(halfway) * halfway;
+
+    return nextDir;
+}
+
+Vector3f MaterialModel::sampleHemisphereCosineWeighted( Vector2f u ) {
+    Vector3f point;
+    
+    Scalar r = std::sqrt(u[0]);
+    Scalar phi = 2 * M_PI * u[1];
+
+    point[0] = r * std::cos(phi);
+    point[1] = r * std::sin(phi);
+    point[2] = std::sqrt(std::max(0.f, 1 - point[0]*point[0] - point[1]*point[1]));
+    
+    return point;
+}
```

Comment:
If we consider that theta is the angle taken from the north pole direction of the hemisphere (whose radius is assumed to be 1), and according to the spherical to cartesian conversion scheme, we have :
```
0 <= theta <= PI/2
0 <= phi <= 2 * PI

x = sin(theta)*cos(phi)
y = sin(theta)*sin(phi)
z = cos(theta)
```
From this, and to prevent computing twice the same information (when sampling and when computing the pdf) and according to https://alexanderameye.github.io/notes/sampling-the-hemisphere/ this method should be
```suggestion
std::pair<Vector3f, Scalar> MaterialModel::sampleHemisphereCosineWeighted( Vector2f u ) {
    Vector3f dir;
    
    Scalar cosTheta = std::sqrt(u[0]);
    Scalar sinTheta = std::sqrt( 1 - u[0] ); // u[0] = cos^2(theta)
    Scalar phi      = 2 * M_PI * u[1];
    
    dir[0] = sinTheta * std::cos(phi);
    dir[1] = sinTheta * std::sin(phi);
    dir[2] = cosTheta
    
    return { dir, cosTheta/M_PI };
}
```

this should be generalized for any _sampleXXX_ method that should return a pair `<sampledDirectio, pdf>`